src/libsyntax/parse/parser.rs

```diff
@@ -515,6 +515,12 @@ impl<'a> Parser<'a> {
         }
     }
 
+    pub fn error_if_typename_is_catch(&mut self, ident: ast::Ident) {
```

Comment:
The goal was to specifically prevent `catch` as a typename so that struct literals and patterns wouldn't conflict with `catch {...}`. Yes, this is a stronger rule than necessary, but if we were to silently interpret `catch { ... }` as a `catch` expression rather than a struct literal/pattern, I think it would cause lots of confusion (especially for code that _used_ to work).

WRT warning against all idents: that seems like far too much of a breaking change. `catch` is a weird name for a type (given Rust's naming conventions), but it's a pretty reasonable name for a variable.